pkg/query/selection_test.go

```diff
@@ -0,0 +1,171 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Selection", func() {
+
+	var ctx context.Context
+	var validCriterion Criterion
+
+	BeforeEach(func() {
+		ctx = context.TODO()
+		validCriterion = ByField(EqualsOperator, "left", "right")
+	})
+
+	Describe("Add criteria to context", func() {
+		addInvalidCriterion := func(criterion Criterion) {
+			ctx, err := AddCriteria(ctx, criterion)
+			Expect(err).To(HaveOccurred())
+			Expect(ctx).To(BeNil())
+		}
+		Context("Invalid", func() {
+			Specify("Univariate operator with multiple right operands", func() {
+				addInvalidCriterion(ByField(EqualsOperator, "leftOp", "1", "2"))
+			})
+			Specify("Nullable operator applied to label query", func() {
+				addInvalidCriterion(ByLabel(EqualsOrNilOperator, "leftOp", "1"))
+			})
+			Specify("Numeric operator to non-numeric right operand", func() {
+				addInvalidCriterion(ByField(GreaterThanOperator, "leftOp", "non-numeric"))
+			})
+			Specify("Field query with duplicate key", func() {
+				var err error
+				ctx, err = AddCriteria(ctx, validCriterion)
+				Expect(err).ToNot(HaveOccurred())
+				addInvalidCriterion(ByField(EqualsOrNilOperator, validCriterion.LeftOp, "right op"))
+			})
+		})
+
+		Context("Valid", func() {
+			Specify("Multivariate operator with single right operand", func() {
+				_, err := AddCriteria(ctx, ByField(InOperator, "leftOp", "1"))
+				Expect(err).ToNot(HaveOccurred())
+			})
+			Specify("With numeric right operand", func() {
+				_, err := AddCriteria(ctx, ByField(LessThanOperator, "leftOp", "5"))
+				Expect(err).ToNot(HaveOccurred())
+			})
+		})
+	})
+
+	Describe("Build criteria from request", func() {
+		var request *web.Request
+		BeforeEach(func() {
+			request = &web.Request{}
+		})
+
+		buildCriteria := func(url string) ([]Criterion, error) {
+			newRequest, err := http.NewRequest(http.MethodGet, url, nil)
+			Expect(err).ToNot(HaveOccurred())
+			request = &web.Request{Request: newRequest}
+			return BuildCriteriaFromRequest(request)
+		}
+
+		Context("When build from request with no query parameters", func() {
+			It("Should return empty criteria", func() {
+				criteriaFromRequest, err := buildCriteria("http://localhost:8080")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(len(criteriaFromRequest)).To(Equal(0))
+			})
+		})
+
+		Context("With missing query operator", func() {
+			It("Should return an error", func() {
+				criteriaFromRequest, err := buildCriteria("http://localhost:8080/v1/visibilities?fieldQuery=leftop_rightop")
+				Expect(err).To(HaveOccurred())
+				Expect(criteriaFromRequest).To(BeNil())
+			})
+		})
+
+		Context("When there is an invalid field query", func() {
+			It("Should return an error", func() {
+				criteriaFromRequest, err := buildCriteria("http://localhost:8080/v1/visibilities?fieldQuery=leftop+lt+rightop")
+				Expect(err).To(HaveOccurred())
+				Expect(criteriaFromRequest).To(BeNil())
+			})
+		})
+
+		Context("When passing multivariate query", func() {
```

Comment:
Same as above  - would it make sense to somehow parameterize threse tests to cover all operations that are defined